### PR TITLE
ARROW-9221: [Java] account for big-endian buffers in ArrowBuf.setBytes

### PR DESCRIPTION
**Pull request by lidavidm**
`ArrowBuf.setBytes` has an override that uses a 8-byte-at-a-time copy loop if the byte buffer does not provide an array and is not direct. Unfortunately, this means it'll mangle data when the byte buffer is big-endian, as it then writes the data into the little-endian ArrowBuf. This fixes it by setting the byte order before copying, and then restoring it.

### CODE REVIEW

**Comment by lidavidm** [comment]
Would a Java maintainer be able to look at this?

**Comment by liyafan82** [comment]
@lidavidm Thanks for reporting the problem.
I am curious how this problem has ever arised. The default byte order is platform dependent. For a big endian machine, the program should crash at the stage of class loading, as our current implementation does not support big endian platforms.

**Comment by lidavidm** [comment]
@liyafan82 The problem actually isn't with big-endian platforms! It's because Java's ByteBuffer [defaults to big-endian](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/ByteBuffer.html#wrap(byte%5B%5D)), and some libraries, like Protobuf, inherit that behavior. We discovered this when developing with Flight internally; there are Protobuf APIs that return a ByteBuffer that may be big-endian; when Flight tries to copy the buffer into an ArrowBuf, it corrupts the data.

**Comment by liyafan82** [comment]
> @liyafan82 The problem actually isn't with big-endian platforms! It's because Java's ByteBuffer [defaults to big-endian](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/ByteBuffer.html#wrap(byte%5B%5D)), and some libraries, like Protobuf, inherit that behavior. We discovered this when developing with Flight internally; there are Protobuf APIs that return a ByteBuffer that may be big-endian; when Flight tries to copy the buffer into an ArrowBuf, it corrupts the data.

@lidavidm Thanks a lot for your clarification. This is really a problem that needs to be fixed.

File: java/memory/src/main/java/org/apache/arrow/memory/ArrowBuf.java [hunk 1]
```diff
@@ -872,6 +874,7 @@ public void setBytes(long index, ByteBuffer src) {
           --length;
           ++dstAddress;
         }
+        src.order(originalByteOrder);
```
Comment (liyafan82) [inline java/memory/src/main/java/org/apache/arrow/memory/ArrowBuf.java:877]: should we place this in a finally block?

Comment (lidavidm) [reply java/memory/src/main/java/org/apache/arrow/memory/ArrowBuf.java]: Good idea, I've updated the PR.

**Review by liyafan82 — APPROVED** [review]
LGTM, thank you.

**Comment by liyafan82** [comment]
Seems a rebase is required.

**Comment by lidavidm** [comment]
Rebased!

**Pull request closed without merging**